src/gui/plugins/mouse_drag/MouseDrag.cc

```diff
@@ -0,0 +1,791 @@
+  }
+  else if (this->dataPtr->mode == MouseDragMode::TRANSLATE)
+  {
+    math::Vector3d origin = this->dataPtr->rayQuery->Origin();
+    if (auto t = this->dataPtr->plane.Intersection(origin, end))
+      this->dataPtr->target = *t;
+    else
+      return;
+
+    math::Vector3d errorPos =
+      this->dataPtr->applicationPoint - this->dataPtr->target;
+    msg = this->dataPtr->CalculateWrench(errorPos, _info.dt, linkWorldPose);
+  }
+
+  // Publish wrench
+  this->dataPtr->pub.Publish(msg);
+}
+
+/////////////////////////////////////////////////
+void MouseDrag::OnSwitchCOM(const bool _checked)
+{
+  this->dataPtr->applyCOM = _checked;
+}
+
+/////////////////////////////////////////////////
+void MouseDragPrivate::OnRender()
+{
+  // Get scene and user camera
+  if (!this->scene)
+  {
+    this->scene = rendering::sceneFromFirstRenderEngine();
+    if (!this->scene)
+    {
+      return;
+    }
+
+    for (unsigned int i = 0; i < this->scene->NodeCount(); ++i)
+    {
+      auto cam = std::dynamic_pointer_cast<rendering::Camera>(
+        this->scene->NodeByIndex(i));
+      if (cam && cam->HasUserData("user-camera") &&
+          std::get<bool>(cam->UserData("user-camera")))
+      {
+        this->camera = cam;
+        gzdbg << "MouseDrag plugin is using camera ["
+              << this->camera->Name() << "]" << std::endl;
+        break;
+      }
+    }
+
+    if (!this->camera)
+    {
+      gzerr << "MouseDrag camera is not available" << std::endl;
+      return;
+    }
+    this->rayQuery = this->scene->CreateRayQuery();
+
+    auto mat = this->scene->Material("Default/TransRed")->Clone();
+    mat->SetDepthCheckEnabled(false);
+    mat->SetDepthWriteEnabled(false);
+
+    this->arrowVisual = this->scene->CreateArrowVisual();
+    this->arrowVisual->SetMaterial(mat);
+    this->arrowVisual->ShowArrowHead(true);
+    this->arrowVisual->ShowArrowShaft(true);
+    this->arrowVisual->ShowArrowRotation(false);
+
+    this->boxVisual = scene->CreateVisual();
+    this->boxVisual->AddGeometry(scene->CreateBox());
+    this->boxVisual->SetInheritScale(false);
+    this->boxVisual->SetMaterial(mat);
+    this->boxVisual->SetUserData("gui-only", static_cast<bool>(true));
+
+    auto planeMat = this->scene->Material("trans-gray");
+    if (!planeMat)
+    {
+      planeMat = this->scene->CreateMaterial("trans-gray");
+      planeMat->SetAmbient(1.0, 1.0, 1.0);
+      planeMat->SetDiffuse(1.0, 1.0, 1.0);
+      planeMat->SetSpecular(1.0, 1.0, 1.0);
+      planeMat->SetTransparency(0.7);
+      planeMat->SetCastShadows(false);
+      planeMat->SetReceiveShadows(false);
+      planeMat->SetLightingEnabled(false);
+    }
+
+    this->planeVisual = scene->CreateVisual();
+    this->planeVisual->AddGeometry(scene->CreatePlane());
+    this->planeVisual->SetInheritScale(false);
+    this->planeVisual->SetMaterial(planeMat);
+    this->planeVisual->SetUserData("gui-only", static_cast<bool>(true));
+  }
+
+  std::lock_guard<std::mutex> lock(this->mutex);
```

Comment:
I think this should be moved up. e.g., `rayQuery` is used in the code above, but it's also used in `Update`.